// million/dist/html.mjs
import { h } from './chunks/h.mjs';

var n=function(t,s,r,e){var u;s[0]=0;for(var h=1;h<s.length;h++){var p=s[h++],a=s[h]?(s[0]|=p?1:2,r[s[h++]]):s[++h];3===p?e[0]=a:4===p?e[1]=Object.assign(e[1]||{},a):5===p?(e[1]=e[1]||{})[s[++h]]=a:6===p?e[1][s[++h]]+=a+"":p?(u=t.apply(a,n(t,a,r,["",null])),e.push(u),a[0]?s[0]|=2:(s[h-2]=0,s[h]=u)):e.push(a);}return e},t=new Map;function htm(s){var r=t.get(this);return r||(r=new Map,t.set(this,r)),(r=n(this,r.get(s)||(r.set(s,r=function(n){for(var t,s,r=1,e="",u="",h=[0],p=function(n){1===r&&(n||(e=e.replace(/^\s*\n\s*|\s*\n\s*$/g,"")))?h.push(0,n,e):3===r&&(n||e)?(h.push(3,n,e),r=2):2===r&&"..."===e&&n?h.push(4,n,0):2===r&&e&&!n?h.push(5,0,!0,e):r>=5&&((e||!n&&5===r)&&(h.push(r,0,e,s),r=6),n&&(h.push(r,n,0,s),r=6)),e="";},a=0;a<n.length;a++){a&&(1===r&&p(),p(a));for(var l=0;l<n[a].length;l++)t=n[a][l],1===r?"<"===t?(p(),h=[h],r=3):e+=t:4===r?"--"===e&&">"===t?(r=1,e=""):e=t+e[0]:u?t===u?u="":e+=t:'"'===t||"'"===t?u=t:">"===t?(p(),r=1):r&&("="===t?(r=5,s=e,e=""):"/"===t&&(r<5||">"===n[a][l+1])?(p(),3===r&&(h=h[0]),r=h,(h=h[0]).push(2,0,r),r=0):" "===t||"\t"===t||"\n"===t||"\r"===t?(p(),r=2):e+=t),3===r&&"!--"===e&&(r=4,h=h[0]);}return p(),h}(s)),r),arguments,[])).length>1?r:r[0]}

const html = htm.bind(h);

const factory = (tagName) => {
  function vnode(param1, param2) {
    if (Array.isArray(param1)) {
      return h(tagName, param2, ...param1);
    } else {
      return h(tagName, param1, ...param2 ?? []);
    }
  }
  return vnode;
};
const input = factory("input");
const textarea = factory("textarea");
const checkbox = factory("input");
const address = factory("address");
const article = factory("article");
const aside = factory("aside");
const footer = factory("footer");
const header = factory("header");
const h1 = factory("h1");
const h2 = factory("h2");
const h3 = factory("h3");
const h4 = factory("h4");
const h5 = factory("h5");
const h6 = factory("h6");
const hgroup = factory("hgroup");
const nav = factory("nav");
const section = factory("section");
const blockquote = factory("blockquote");
const dd = factory("dd");
const div = factory("div");
const dl = factory("dl");
const dt = factory("dt");
const figcaption = factory("figcaption");
const figure = factory("figure");
const hr = factory("hr");
const li = factory("li");
const main = factory("main");
const ol = factory("ol");
const p = factory("p");
const pre = factory("pre");
const ul = factory("ul");
const a = factory("a");
const abbr = factory("abbr");
const b = factory("b");
const bdi = factory("bdi");
const bdo = factory("bdo");
const br = factory("br");
const cite = factory("cite");
const code = factory("code");
const data = factory("data");
const dfn = factory("dfn");
const em = factory("em");
const i = factory("i");
const kbd = factory("kbd");
const mark = factory("mark");
const q = factory("q");
const rp = factory("rp");
const rt = factory("rt");
const rtc = factory("rtc");
const ruby = factory("ruby");
const s = factory("s");
const samp = factory("samp");
const small = factory("small");
const span = factory("span");
const strong = factory("strong");
const sub = factory("sub");
const sup = factory("sup");
const time = factory("time");
const u = factory("u");
const varElement = factory("var");
const wbr = factory("wbr");
const area = factory("area");
const audio = factory("audio");
const img = factory("img");
const map = factory("map");
const track = factory("track");
const video = factory("video");
const embed = factory("embed");
const object = factory("object");
const param = factory("param");
const picture = factory("picture");
const source = factory("source");
const canvas = factory("canvas");
const script = factory("script");
const del = factory("del");
const ins = factory("ins");
const caption = factory("caption");
const col = factory("col");
const colgroup = factory("colgroup");
const table = factory("table");
const tbody = factory("tbody");
const td = factory("td");
const tfoot = factory("tfoot");
const th = factory("th");
const thead = factory("thead");
const tr = factory("tr");
const button = factory("button");
const datalist = factory("datalist");
const fieldset = factory("fieldset");
const form = factory("form");
const label = factory("label");
const legend = factory("legend");
const meter = factory("meter");
const optgroup = factory("optgroup");
const option = factory("option");
const output = factory("output");
const progress = factory("progress");
const select = factory("select");
const details = factory("details");
const menuitem = factory("menuitem");
const summary = factory("summary");
const slot = factory("slot");
const template = factory("template");
const circle = factory("circle");
const rect = factory("rect");
const ellipse = factory("ellipse");
const g = factory("g");
const image = factory("image");
const line = factory("line");
const mask = factory("mask");
const path = factory("path");
const polygon = factory("polygon");
const polyline = factory("polyline");
const svg = factory("svg");
const svgText = factory("text");
const marker = factory("marker");
const linearGradient = factory("linearGradient");
const foreignObject = factory("foreignObject");

export { a, abbr, address, area, article, aside, audio, b, bdi, bdo, blockquote, br, button, canvas, caption, checkbox, circle, cite, code, col, colgroup, data, datalist, dd, del, details, dfn, div, dl, dt, ellipse, em, embed, factory, fieldset, figcaption, figure, footer, foreignObject, form, g, h1, h2, h3, h4, h5, h6, header, hgroup, hr, html, i, image, img, input, ins, kbd, label, legend, li, line, linearGradient, main, map, mark, marker, mask, menuitem, meter, nav, object, ol, optgroup, option, output, p, param, path, picture, polygon, polyline, pre, progress, q, rect, rp, rt, rtc, ruby, s, samp, script, section, select, slot, small, source, span, strong, sub, summary, sup, svg, svgText, table, tbody, td, template, textarea, tfoot, th, thead, time, tr, track, u, ul, varElement, video, wbr };
