// million/dist/million.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const render = require('./chunks/render.cjs');
const h = require('./chunks/h.cjs');
require('./chunks/convert.cjs');

const _ = void 0;

exports.createElement = render.createElement;
exports.diff = render.diff;
exports.flushQueue = render.flushQueue;
exports.hydrate = render.hydrate;
exports.isPending = render.isPending;
exports.patch = render.patch;
exports.render = render.render;
exports.startTransition = render.startTransition;
exports.updateProp = render.updateProp;
exports.useChildren = render.useChildren;
exports.useNode = render.useNode;
exports.useProps = render.useProps;
exports.DeltaTypes = h.DeltaTypes;
exports.Deltas = h.Deltas;
exports.EffectTypes = h.EffectTypes;
exports.Flags = h.Flags;
exports.OLD_VNODE_FIELD = h.OLD_VNODE_FIELD;
exports.className = h.className;
exports.entity = h.entity;
exports.kebab = h.kebab;
exports.m = h.m;
exports.ns = h.ns;
exports.resolveVNode = h.resolveVNode;
exports.style = h.style;
exports.svg = h.svg;
exports._ = _;
