// million/dist/ssr.cjs
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

const morph = require('./chunks/morph.cjs');
const convert = require('./chunks/convert.cjs');
const h = require('./chunks/h.cjs');
require('./chunks/render.cjs');

const refresh = (head, body) => {
  if (head)
    morph.morph(head, document.head);
  if (body)
    morph.morph(body, document.body);
};

const renderToString = (vnode) => {
  if (typeof vnode === "string")
    return vnode;
  if (vnode === void 0)
    return "<!-- -->";
  let attributes = "";
  let children = "";
  for (const prop in vnode.props) {
    if (!prop.toLowerCase().startsWith("on")) {
      attributes += ` ${prop}="${vnode.props[prop]}"`;
    }
  }
  if (htmlVoidElements.includes(vnode.tag)) {
    return `<${vnode.tag}${attributes} />`;
  }
  for (const child of vnode.children || []) {
    children += renderToString(child);
  }
  return `<${vnode.tag}${attributes}>${children}</${vnode.tag}>`;
};
const htmlVoidElements = [
  "area",
  "base",
  "basefont",
  "bgsound",
  "br",
  "col",
  "command",
  "embed",
  "frame",
  "hr",
  "image",
  "img",
  "input",
  "isindex",
  "keygen",
  "link",
  "menuitem",
  "meta",
  "nextid",
  "param",
  "source",
  "track",
  "wbr"
];

exports.fromDomNodeToVNode = convert.fromDomNodeToVNode;
exports.fromStringToDomNode = convert.fromStringToDomNode;
exports.h = h.h;
exports.normalize = h.normalize;
exports.htmlVoidElements = htmlVoidElements;
exports.refresh = refresh;
exports.renderToString = renderToString;
